### ISSUE TITLE
Python underlinking issue (-Wl,--no-undefined)

### ISSUE THREAD

**Issue by akien-mga** [baldurk/renderdoc#639]
When building renderdoc 0.34 with the following linkflags, an [underlinking issue](https://wiki.mageia.org/en/Underlinking_issues_in_packaging) appears for the SWIG bindings:
```
LDFLAGS="-Wl,--as-needed -Wl,--no-undefined -Wl,-z,relro -Wl,-O1 -Wl,--build-id -Wl,--enable-new-dtags"
```
```
[100%] Linking CXX shared library ../../../bin/_qrenderdoc.so
cd /home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/Code/pyrenderdoc && /usr/bin/cmake -E cmake_link_script CMakeFiles/_qrenderdoc.dir/link.txt --verbose=1
/bin/c++  -fPIC -I/usr/lib64/qt5/include/QtWidgets -I/usr/lib64/qt5/include -I/usr/lib64/qt5/include/QtGui -I/usr/lib64/qt5/include -I/usr/lib64/qt5/include/QtCore -I/usr/lib64/qt5/include -O2 -g -pipe -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4 -std=c++11 -fstrict-aliasing -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wextra -Wno-unused-variable -Wno-unused-parameter -Wno-unused-result -Wno-type-limits -Wno-missing-field-initializers -Wno-unknown-pragmas -Wno-reorder -Wno-unused-but-set-variable -Werror -DNDEBUG -lQt5Widgets -lQt5Gui -lQt5Core -Wl,--as-needed -Wl,--no-undefined -Wl,-z,relro -Wl,-O1 -Wl,--build-id -Wl,--enable-new-dtags -shared -Wl,-soname,_qrenderdoc.so -o ../../../bin/_qrenderdoc.so CMakeFiles/_qrenderdoc.dir/__/__/qrenderdoc_python.cxx.o CMakeFiles/_qrenderdoc.dir/__/Interface/CommonPipelineState.cpp.o CMakeFiles/_qrenderdoc.dir/__/Interface/PersistantConfig.cpp.o CMakeFiles/_qrenderdoc.dir/__/Interface/QRDInterface.cpp.o CMakeFiles/_qrenderdoc.dir/__/Interface/RemoteHost.cpp.o CMakeFiles/_qrenderdoc.dir/pyrenderdoc_stub.cpp.o CMakeFiles/_qrenderdoc.dir/qrenderdoc_stub.cpp.o ../../../bin/librenderdoc.so -lQt5Widgets -lQt5Gui -lQt5Core 
CMakeFiles/_qrenderdoc.dir/__/__/qrenderdoc_python.cxx.o: In function `SWIG_Python_ErrorType(int)':
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1031: undefined reference to `PyExc_RuntimeError'
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1001: undefined reference to `PyExc_IOError'
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:998: undefined reference to `PyExc_MemoryError'
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1028: undefined reference to `PyExc_AttributeError'
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1025: undefined reference to `PyExc_SystemError'
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1022: undefined reference to `PyExc_ValueError'
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1019: undefined reference to `PyExc_SyntaxError'
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1016: undefined reference to `PyExc_OverflowError'
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1013: undefined reference to `PyExc_ZeroDivisionError'
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1010: undefined reference to `PyExc_TypeError'
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1007: undefined reference to `PyExc_IndexError'
CMakeFiles/_qrenderdoc.dir/__/__/qrenderdoc_python.cxx.o: In function `SwigPyObject_disown':
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1453: undefined reference to `_Py_NoneStruct'
CMakeFiles/_qrenderdoc.dir/__/__/qrenderdoc_python.cxx.o: In function `SwigPyObject_acquire':
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1453: undefined reference to `_Py_NoneStruct'
CMakeFiles/_qrenderdoc.dir/__/__/qrenderdoc_python.cxx.o: In function `SwigPyObject_get___dict__':
/home/akien/Mageia/Checkout/renderdoc/BUILD/renderdoc-0.34/build/qrenderdoc/qrenderdoc_python.cxx:1567: undefined reference to `PyDict_New'
[snip]
```
I assume those symbols are provided by the custom built swig, so it should likely be in the linker command.

Related question: would it be possible to allow unbundling swig in a future release? (i.e. build against the system library instead of your modified one - albeit with loss of functionality I suppose)

**Comment by baldurk** [comment]
This doesn't look related to the custom SWIG build to me, at this point SWIG has already successfully compiled and linked and been used to generate the bindings. SWIG only generates binding code that is self-contained, you don't ever link against SWIG as part of the build - those missing symbols come from  python itself.

This can be fixed - I'll set those python modules I generate (that's one there) to explicitly link against python, but I was hoping to avoid doing that as it gives up compatibility since you're then fixing a particular major.minor python version. Python's module system seems like a bit of a mess in that respect though.

There's no feasible way to use the system library, it's not a matter of dropping functionality as the system SWIG would just fail to parse the code and on top of that produce an incompatibly different python API. You'd have to drop the python integration entirely which I don't really want to make conditional.

**Comment by baldurk** [comment]
I've linked against python for now, can you check if that's working?

FYI in future if I start including these modules as part of the distribution (... somehow. No idea how or if it's a good idea) instead of just a build artifact for generating docs, then this change might get reverted so that the modules work across python versions.

**Comment by akien-mga** [comment]
Thanks, I'll give it a try tonight.

>FYI in future if I start including these modules as part of the distribution (... somehow. No idea how or if it's a good idea) instead of just a build artifact for generating docs, then this change might get reverted so that the modules work across python versions.

Yeah, that makes sense. Maybe you could then make it opt-in, so that distros could still enforce proper linking - if python gets updated in the distro, all reverse dependencies are rebuilt too, so for Linux distros that should not impact your use case as upstream distributor.

**Comment by akien-mga** [comment]
Confirmed fixed by ffcbf23.
